Transaction fee and change calculation failed... using v0.21.0 binary

After upgrade of binary bitcoind from 0.19.1 to 0.20.0 could not process transactions. Did not see anything in release notes that would be breaking. Any idea what could be the issue? I cant find any documentation on this error message.

**Expected behavior**

Process transaction without error using "sendtoaddress"

**Actual behavior**

error code -4 "Transaction fee and change calculation failed"

**To reproduce**

All we did was upgrade binary bitcoind from 0.19.1 to v0.20.0. Nothing else changed.

**System information**

Using binary from https://bitcoincore.org/bin/bitcoin-core-0.20.0/bitcoin-0.20.0-x86_64-linux-gnu.tar.gz

On Linux Ubuntu 18.04.3 LTS on a SSD

No ERRORs exist in debug.log. Node syncs fine on v0.20.0... The following seems relevant but we saw same logs on v0.19.1.

**v0.19.1...**

2020-06-09T07:51:36Z [default wallet] Fee Calculation: Fee:2871 Bytes:136 Needed:2871 Tgt:6 (requested 6) Reason:"Conservative Double Target longer horizon" Decay 0.99931: Estimation: (20593.8 - 21623.5) 97.20% 37504.3/(38582.5 0 mem 3.0 out) Fail: (19613.1 - 20593.8) 92.07% 24792.5/(26908.8 0 mem 19.2 out)
2020-06-09T07:54:32Z [default wallet] Fee Calculation: Fee:13110 Bytes:621 Needed:13110 Tgt:6 (requested 6) Reason:"Conservative Double Target longer horizon" Decay 0.99931: Estimation: (20593.8 - 21623.5) 97.20% 37504.3/(38582.5 0 mem 3.0 out) Fail: (19613.1 - 20593.8) 92.07% 24792.5/(26908.8 0 mem 19.2 out)
2020-06-09T09:58:09Z [default wallet] Fee Calculation: Fee:7326 Bytes:347 Needed:7326 Tgt:6 (requested 6) Reason:"Conservative Double Target longer horizon" Decay 0.99931: Estimation: (20593.8 - 21623.5) 97.39% 40008.9/(41076.7 0 mem 3.0 out) Fail: (19613.1 - 20593.8) 92.28% 25270.9/(27366.8 0 mem 19.0 out)
2020-06-09T13:22:31Z [default wallet] Fee Calculation: Fee:16956 Bytes:803 Needed:16956 Tgt:6 (requested 6) Reason:"Conservative Double Target longer horizon" Decay 0.99931: Estimation: (20593.8 - 21623.5) 97.64% 43671.2/(44725.9 0 mem 3.0 out) Fail: (19613.1 - 20593.8) 92.48% 25689.0/(27759.1 0 mem 18.8 out)
2020-06-09T14:36:45Z [default wallet] Fee Calculation: Fee:54877 Bytes:2255 Needed:54877 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (23839.9 - 25031.9) 98.23% 3643.5/(3709.2 0 mem 0.0 out) Fail: (22704.7 - 23839.9) 81.84% 2485.0/(3036.6 0 mem 0.0 out)
2020-06-09T15:33:30Z [default wallet] Fee Calculation: Fee:5709 Bytes:258 Needed:5709 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (21623.5 - 22704.7) 88.84% 3368.3/(3791.3 0 mem 0.0 out) Fail: (20593.8 - 21623.5) 77.45% 3045.0/(3928.4 0 mem 3.2 out)
2020-06-09T16:06:13Z [default wallet] Fee Calculation: Fee:90020 Bytes:4068 Needed:90020 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (21623.5 - 22704.7) 88.87% 3127.1/(3518.6 0 mem 0.0 out) Fail: (20593.8 - 21623.5) 77.48% 2822.9/(3640.5 0 mem 2.9 out)
2020-06-09T18:16:54Z [default wallet] Fee Calculation: Fee:7683 Bytes:347 Needed:7683 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (21623.5 - 22704.7) 92.06% 3166.4/(3439.3 0 mem 0.0 out) Fail: (20593.8 - 21623.5) 82.08% 2307.7/(2809.5 0 mem 2.1 out)
2020-06-09T18:28:31Z [default wallet] Fee Calculation: Fee:11759 Bytes:531 Needed:11759 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (21623.5 - 22704.7) 92.58% 3152.3/(3404.9 0 mem 0.0 out) Fail: (20593.8 - 21623.5) 82.64% 2219.5/(2683.9 0 mem 2.0 out)
2020-06-09T18:53:50Z [default wallet] Fee Calculation: Fee:11716 Bytes:529 Needed:11716 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (21623.5 - 22704.7) 93.28% 3003.3/(3219.7 0 mem 0.0 out) Fail: (20593.8 - 21623.5) 83.47% 2016.6/(2414.3 0 mem 1.7 out)
2020-06-09T19:18:43Z [default wallet] Fee Calculation: Fee:4962 Bytes:224 Needed:4962 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (21623.5 - 22704.7) 93.59% 2925.7/(3126.0 0 mem 0.0 out) Fail: (20593.8 - 21623.5) 83.74% 1903.9/(2272.0 0 mem 1.6 out)

**v0.20.0 ...**

2020-06-09T20:18:06Z [default wallet] Fee Calculation: Fee:2829 Bytes:134 Needed:2829 Tgt:6 (requested 6) Reason:"Conservative Double Target longer horizon" Decay 0.99931: Estimation: (20593.8 - 21623.5) 97.79% 45111.8/(46130.7 0 mem 2.9 out) Fail: (19613.1 - 20593.8) 93.00% 26822.9/(28822.8 0 mem 18.2 out)

**v0.19.1...**

2020-06-09T21:44:39Z [default wallet] Fee Calculation: Fee:7372 Bytes:349 Needed:7372 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (20593.8 - 21623.5) 91.12% 2121.8/(2327.6 0 mem 0.9 out) Fail: (19613.1 - 20593.8) 83.14% 777.5/(935.0 0 mem 0.1 out)
2020-06-09T21:45:57Z [default wallet] Fee Calculation: Fee:7330 Bytes:347 Needed:7330 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (20593.8 - 21623.5) 91.12% 2121.8/(2327.6 0 mem 0.9 out) Fail: (19613.1 - 20593.8) 83.14% 777.5/(935.0 0 mem 0.1 out)
2020-06-09T21:50:48Z [default wallet] Fee Calculation: Fee:36125 Bytes:1710 Needed:36125 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (20593.8 - 21623.5) 91.12% 2121.8/(2327.6 0 mem 0.9 out) Fail: (19613.1 - 20593.8) 83.14% 777.5/(935.0 0 mem 0.1 out)
2020-06-09T22:04:02Z [default wallet] Fee Calculation: Fee:4775 Bytes:226 Needed:4775 Tgt:6 (requested 6) Reason:"Target 85% Threshold" Decay 0.96200: Estimation: (20593.8 - 21623.5) 92.53% 2369.9/(2560.4 0 mem 0.8 out) Fail: (19613.1 - 20593.8) 84.30% 783.4/(929.1 0 mem 0.1 out)

> Hi @SomethingUseful you can build from source at the head commit of #17331 (see the `doc/build-*` guides or [this tutorial](https://jonatack.github.io/articles/how-to-compile-bitcoin-core-and-run-the-tests) if needed) and test or run bitcoin core with that (with small amounts at first) and leave review feedback if it resolves the issue for you.

Tried this and it errored... anything obvious I am missing?

STEPS TO COMPILE w/ PR:
```
# apt update && apt dist-upgrade
# apt-get install build-essential libtool autotools-dev automake pkg-config bsdmainutils python3 libssl-dev libevent-dev libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-test-dev libboost-thread-dev libminiupnpc-dev libzmq3-dev libqt5gui5 libqt5core5a libqt5dbus5 qttools5-dev qttools5-dev-tools libprotobuf-dev protobuf-compiler git libsqlite3-dev ccache
# cd /usr/local/src
# mkdir bitcoin-0.21.0-patch
# cd bitcoin-0.21.0-patch/
# git clone https://github.com/bitcoin/bitcoin.git .
# ./contrib/install_db4.sh `pwd`

# nano .git/config
fetch = +refs/pull/*/head:refs/remotes/origin/pr/*

# git fetch origin
# git checkout pr/17331
# export BDB_PREFIX='/usr/local/src/bitcoin-0.21.0-patch/db4'
# ./autogen.sh
# ./configure BDB_LIBS="-L${BDB_PREFIX}/lib -ldb_cxx-4.8" BDB_CFLAGS="-I${BDB_PREFIX}/include"
# make -j "$(($(nproc)+1))"
```
 
  CXX      wallet/libbitcoin_wallet_a-rpcwallet.o
g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
Makefile:8894: recipe for target 'wallet/libbitcoin_wallet_a-rpcwallet.o' failed
make[2]: *** [wallet/libbitcoin_wallet_a-rpcwallet.o] Error 4
make[2]: Leaving directory '/usr/local/src/bitcoin-0.21.0-patch/src'
Makefile:14461: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/usr/local/src/bitcoin-0.21.0-patch/src'
Makefile:812: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1

--- a/src/bench/coin_selection.cpp
+++ b/src/bench/coin_selection.cpp
@@ -50,7 +50,7 @@ static void CoinSelection(benchmark::Bench& bench)
 
     const CoinEligibilityFilter filter_standard(1, 6, 0);
     const CoinSelectionParams coin_selection_params(/* use_bnb= */ true, /* change_output_size= */ 34,
-                                                    /* change_spend_size= */ 148, /* effective_fee= */ CFeeRate(0),
+                                                    /* change_spend_size= */ 148, /* effective_feerate= */ CFeeRate(0),
                                                     /* long_term_feerate= */ CFeeRate(0), /* discard_feerate= */ CFeeRate(0),
                                                     /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
     bench.run([&] {

--- a/src/bench/coin_selection.cpp
+++ b/src/bench/coin_selection.cpp
@@ -51,7 +51,7 @@ static void CoinSelection(benchmark::Bench& bench)
     const CoinEligibilityFilter filter_standard(1, 6, 0);
     const CoinSelectionParams coin_selection_params(/* use_bnb= */ true, /* change_output_size= */ 34,
                                                     /* change_spend_size= */ 148, /* effective_fee= */ CFeeRate(0),
-                                                    /* long_term_feerate= */ CFeeRate(0),
+                                                    /* long_term_feerate= */ CFeeRate(0), /* discard_feerate= */ CFeeRate(0),
                                                     /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
     bench.run([&] {
         std::set<CInputCoin> setCoinsRet;

--- a/src/bench/coin_selection.cpp
+++ b/src/bench/coin_selection.cpp
@@ -49,7 +49,10 @@ static void CoinSelection(benchmark::Bench& bench)
     }
 
     const CoinEligibilityFilter filter_standard(1, 6, 0);
-    const CoinSelectionParams coin_selection_params(true, 34, 148, CFeeRate(0), 0, false);
+    const CoinSelectionParams coin_selection_params(/* use_bnb= */ true, /* change_output_size= */ 34,
+                                                    /* change_spend_size= */ 148, /* effective_fee= */ CFeeRate(0),
+                                                    /* long_term_feerate= */ CFeeRate(0),
+                                                    /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
     bench.run([&] {
         std::set<CInputCoin> setCoinsRet;
         CAmount nValueRet;

--- a/src/wallet/test/coinselector_tests.cpp
+++ b/src/wallet/test/coinselector_tests.cpp
@@ -37,7 +37,7 @@ CoinEligibilityFilter filter_confirmed(1, 1, 0);
 CoinEligibilityFilter filter_standard_extra(6, 6, 0);
 CoinSelectionParams coin_selection_params(/* use_bnb= */ false, /* change_output_size= */ 0,
                                           /* change_spend_size= */ 0, /* effective_fee= */ CFeeRate(0),
-                                          /* long_term_feerate= */ CFeeRate(0),
+                                          /* long_term_feerate= */ CFeeRate(0), /* discard_feerate= */ CFeeRate(0),
                                           /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
 
 static void add_coin(const CAmount& nValue, int nInput, std::vector<CInputCoin>& set)
@@ -274,7 +274,7 @@ BOOST_AUTO_TEST_CASE(bnb_search_test)
     // Make sure that effective value is working in SelectCoinsMinConf when BnB is used
     CoinSelectionParams coin_selection_params_bnb(/* use_bnb= */ true, /* change_output_size= */ 0,
                                                   /* change_spend_size= */ 0, /* effective_fee= */ CFeeRate(3000),
-                                                  /* long_term_feerate= */ CFeeRate(1000),
+                                                  /* long_term_feerate= */ CFeeRate(1000), /* discard_feerate= */ CFeeRate(1000),
                                                   /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
     CoinSet setCoinsRet;
     CAmount nValueRet;
@@ -647,11 +647,11 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
         // Perform selection
         CoinSelectionParams coin_selection_params_knapsack(/* use_bnb= */ false, /* change_output_size= */ 34,
                                                            /* change_spend_size= */ 148, /* effective_fee= */ CFeeRate(0),
-                                                           /* long_term_feerate= */ CFeeRate(0),
+                                                           /* long_term_feerate= */ CFeeRate(0), /* discard_feerate= */ CFeeRate(0),
                                                            /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
         CoinSelectionParams coin_selection_params_bnb(/* use_bnb= */ true, /* change_output_size= */ 34,
                                                       /* change_spend_size= */ 148, /* effective_fee= */ CFeeRate(0),
-                                                      /* long_term_feerate= */ CFeeRate(0),
+                                                      /* long_term_feerate= */ CFeeRate(0), /* discard_feerate= */ CFeeRate(0),
                                                       /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
         CoinSet out_set;
         CAmount out_value = 0;

--- a/src/wallet/test/coinselector_tests.cpp
+++ b/src/wallet/test/coinselector_tests.cpp
@@ -35,7 +35,10 @@ static CAmount balance = 0;
 CoinEligibilityFilter filter_standard(1, 6, 0);
 CoinEligibilityFilter filter_confirmed(1, 1, 0);
 CoinEligibilityFilter filter_standard_extra(6, 6, 0);
-CoinSelectionParams coin_selection_params(false, 0, 0, CFeeRate(0), 0, false);
+CoinSelectionParams coin_selection_params(/* use_bnb= */ false, /* change_output_size= */ 0,
+                                          /* change_spend_size= */ 0, /* effective_fee= */ CFeeRate(0),
+                                          /* long_term_feerate= */ CFeeRate(0),
+                                          /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
 
 static void add_coin(const CAmount& nValue, int nInput, std::vector<CInputCoin>& set)
 {
@@ -269,7 +272,10 @@ BOOST_AUTO_TEST_CASE(bnb_search_test)
     }
 
     // Make sure that effective value is working in SelectCoinsMinConf when BnB is used
-    CoinSelectionParams coin_selection_params_bnb(true, 0, 0, CFeeRate(3000), 0, false);
+    CoinSelectionParams coin_selection_params_bnb(/* use_bnb= */ true, /* change_output_size= */ 0,
+                                                  /* change_spend_size= */ 0, /* effective_fee= */ CFeeRate(3000),
+                                                  /* long_term_feerate= */ CFeeRate(1000),
+                                                  /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
     CoinSet setCoinsRet;
     CAmount nValueRet;
     bool bnb_used;
@@ -639,8 +645,14 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
         CAmount target = rand.randrange(balance - 1000) + 1000;
 
         // Perform selection
-        CoinSelectionParams coin_selection_params_knapsack(false, 34, 148, CFeeRate(0), 0, false);
-        CoinSelectionParams coin_selection_params_bnb(true, 34, 148, CFeeRate(0), 0, false);
+        CoinSelectionParams coin_selection_params_knapsack(/* use_bnb= */ false, /* change_output_size= */ 34,
+                                                           /* change_spend_size= */ 148, /* effective_fee= */ CFeeRate(0),
+                                                           /* long_term_feerate= */ CFeeRate(0),
+                                                           /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
+        CoinSelectionParams coin_selection_params_bnb(/* use_bnb= */ true, /* change_output_size= */ 34,
+                                                      /* change_spend_size= */ 148, /* effective_fee= */ CFeeRate(0),
+                                                      /* long_term_feerate= */ CFeeRate(0),
+                                                      /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
         CoinSet out_set;
         CAmount out_value = 0;
         bool bnb_used = false;

--- a/src/wallet/test/coinselector_tests.cpp
+++ b/src/wallet/test/coinselector_tests.cpp
@@ -36,7 +36,7 @@ CoinEligibilityFilter filter_standard(1, 6, 0);
 CoinEligibilityFilter filter_confirmed(1, 1, 0);
 CoinEligibilityFilter filter_standard_extra(6, 6, 0);
 CoinSelectionParams coin_selection_params(/* use_bnb= */ false, /* change_output_size= */ 0,
-                                          /* change_spend_size= */ 0, /* effective_fee= */ CFeeRate(0),
+                                          /* change_spend_size= */ 0, /* effective_feerate= */ CFeeRate(0),
                                           /* long_term_feerate= */ CFeeRate(0), /* discard_feerate= */ CFeeRate(0),
                                           /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
 
@@ -273,7 +273,7 @@ BOOST_AUTO_TEST_CASE(bnb_search_test)
 
     // Make sure that effective value is working in SelectCoinsMinConf when BnB is used
     CoinSelectionParams coin_selection_params_bnb(/* use_bnb= */ true, /* change_output_size= */ 0,
-                                                  /* change_spend_size= */ 0, /* effective_fee= */ CFeeRate(3000),
+                                                  /* change_spend_size= */ 0, /* effective_feerate= */ CFeeRate(3000),
                                                   /* long_term_feerate= */ CFeeRate(1000), /* discard_feerate= */ CFeeRate(1000),
                                                   /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
     CoinSet setCoinsRet;
@@ -307,7 +307,7 @@ BOOST_AUTO_TEST_CASE(bnb_search_test)
         CCoinControl coin_control;
         coin_control.fAllowOtherInputs = true;
         coin_control.Select(COutPoint(vCoins.at(0).tx->GetHash(), vCoins.at(0).i));
-        coin_selection_params_bnb.effective_fee = CFeeRate(0);
+        coin_selection_params_bnb.m_effective_feerate = CFeeRate(0);
         BOOST_CHECK(wallet->SelectCoins(vCoins, 10 * CENT, setCoinsRet, nValueRet, coin_control, coin_selection_params_bnb, bnb_used));
         BOOST_CHECK(bnb_used);
         BOOST_CHECK(coin_selection_params_bnb.use_bnb);
@@ -646,11 +646,11 @@ BOOST_AUTO_TEST_CASE(SelectCoins_test)
 
         // Perform selection
         CoinSelectionParams coin_selection_params_knapsack(/* use_bnb= */ false, /* change_output_size= */ 34,
-                                                           /* change_spend_size= */ 148, /* effective_fee= */ CFeeRate(0),
+                                                           /* change_spend_size= */ 148, /* effective_feerate= */ CFeeRate(0),
                                                            /* long_term_feerate= */ CFeeRate(0), /* discard_feerate= */ CFeeRate(0),
                                                            /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
         CoinSelectionParams coin_selection_params_bnb(/* use_bnb= */ true, /* change_output_size= */ 34,
-                                                      /* change_spend_size= */ 148, /* effective_fee= */ CFeeRate(0),
+                                                      /* change_spend_size= */ 148, /* effective_feerate= */ CFeeRate(0),
                                                       /* long_term_feerate= */ CFeeRate(0), /* discard_feerate= */ CFeeRate(0),
                                                       /* tx_no_inputs_size= */ 0, /* avoid_partial= */ false);
         CoinSet out_set;

--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2374,7 +2374,7 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
         std::vector<OutputGroup> groups = GroupOutputs(coins, !coin_selection_params.m_avoid_partial_spends, effective_feerate, coin_selection_params.m_long_term_feerate, eligibility_filter, true /* positive_only */);
 
         // Calculate cost of change
-        CAmount cost_of_change = GetDiscardRate(*this).GetFee(coin_selection_params.change_spend_size) + coin_selection_params.effective_fee.GetFee(coin_selection_params.change_output_size);
+        CAmount cost_of_change = coin_selection_params.m_discard_feerate.GetFee(coin_selection_params.change_spend_size) + coin_selection_params.effective_fee.GetFee(coin_selection_params.change_output_size);
 
         // Calculate the fees for things that aren't inputs
         CAmount not_input_fees = coin_selection_params.effective_fee.GetFee(coin_selection_params.tx_noinputs_size);
@@ -2798,7 +2798,8 @@ bool CWallet::CreateTransactionInternal(
             CTxOut change_prototype_txout(0, scriptChange);
             coin_selection_params.change_output_size = GetSerializeSize(change_prototype_txout);
 
-            CFeeRate discard_rate = GetDiscardRate(*this);
+            // Set discard feerate
+            coin_selection_params.m_discard_feerate = GetDiscardRate(*this);
 
             // Get the fee rate to use effective values in coin selection
             coin_selection_params.effective_fee = GetMinimumFeeRate(*this, coin_control, &feeCalc);
@@ -2917,7 +2918,7 @@ bool CWallet::CreateTransactionInternal(
                     // Never create dust outputs; if we would, just
                     // add the dust to the fee.
                     // The nChange when BnB is used is always going to go to fees.
-                    if (IsDust(newTxOut, discard_rate) || bnb_used)
+                    if (IsDust(newTxOut, coin_selection_params.m_discard_feerate) || bnb_used)
                     {
                         nChangePosInOut = -1;
                         nFeeRet += nChange;
@@ -2969,7 +2970,7 @@ bool CWallet::CreateTransactionInternal(
                     if (nChangePosInOut == -1 && nSubtractFeeFromAmount == 0 && pick_new_inputs) {
                         unsigned int tx_size_with_change = nBytes + coin_selection_params.change_output_size + 2; // Add 2 as a buffer in case increasing # of outputs changes compact size
                         CAmount fee_needed_with_change = coin_selection_params.effective_fee.GetFee(tx_size_with_change);
-                        CAmount minimum_value_for_change = GetDustThreshold(change_prototype_txout, discard_rate);
+                        CAmount minimum_value_for_change = GetDustThreshold(change_prototype_txout, coin_selection_params.m_discard_feerate);
                         if (nFeeRet >= fee_needed_with_change + minimum_value_for_change) {
                             pick_new_inputs = false;
                             nFeeRet = fee_needed_with_change;

--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2364,12 +2364,6 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
     nValueRet = 0;
 
     if (coin_selection_params.use_bnb) {
-        // Get long term estimate
-        FeeCalculation feeCalc;
-        CCoinControl temp;
-        temp.m_confirm_target = 1008;
-        CFeeRate long_term_feerate = GetMinimumFeeRate(*this, temp, &feeCalc);
-
         // Get the feerate for effective value.
         // When subtracting the fee from the outputs, we want the effective feerate to be 0
         CFeeRate effective_feerate{0};
@@ -2377,7 +2371,7 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
             effective_feerate = coin_selection_params.effective_fee;
         }
 
-        std::vector<OutputGroup> groups = GroupOutputs(coins, !coin_selection_params.m_avoid_partial_spends, effective_feerate, long_term_feerate, eligibility_filter, true /* positive_only */);
+        std::vector<OutputGroup> groups = GroupOutputs(coins, !coin_selection_params.m_avoid_partial_spends, effective_feerate, coin_selection_params.m_long_term_feerate, eligibility_filter, true /* positive_only */);
 
         // Calculate cost of change
         CAmount cost_of_change = GetDiscardRate(*this).GetFee(coin_selection_params.change_spend_size) + coin_selection_params.effective_fee.GetFee(coin_selection_params.change_output_size);
@@ -2819,6 +2813,11 @@ bool CWallet::CreateTransactionInternal(
                 error = _("Fee estimation failed. Fallbackfee is disabled. Wait a few blocks or enable -fallbackfee.");
                 return false;
             }
+
+            // Get long term estimate
+            CCoinControl cc_temp;
+            cc_temp.m_confirm_target = chain().estimateMaxBlocks();
+            coin_selection_params.m_long_term_feerate = GetMinimumFeeRate(*this, cc_temp, nullptr);
 
             nFeeRet = 0;
             bool pick_new_inputs = true;

--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2807,11 +2807,11 @@ bool CWallet::CreateTransactionInternal(
             CFeeRate discard_rate = GetDiscardRate(*this);
 
             // Get the fee rate to use effective values in coin selection
-            CFeeRate nFeeRateNeeded = GetMinimumFeeRate(*this, coin_control, &feeCalc);
+            coin_selection_params.effective_fee = GetMinimumFeeRate(*this, coin_control, &feeCalc);
             // Do not, ever, assume that it's fine to change the fee rate if the user has explicitly
             // provided one
-            if (coin_control.m_feerate && nFeeRateNeeded > *coin_control.m_feerate) {
-                error = strprintf(_("Fee rate (%s) is lower than the minimum fee rate setting (%s)"), coin_control.m_feerate->ToString(FeeEstimateMode::SAT_VB), nFeeRateNeeded.ToString(FeeEstimateMode::SAT_VB));
+            if (coin_control.m_feerate && coin_selection_params.effective_fee > *coin_control.m_feerate) {
+                error = strprintf(_("Fee rate (%s) is lower than the minimum fee rate setting (%s)"), coin_control.m_feerate->ToString(FeeEstimateMode::SAT_VB), coin_selection_params.effective_fee.ToString(FeeEstimateMode::SAT_VB));
                 return false;
             }
             if (feeCalc.reason == FeeReason::FALLBACK && !m_allow_fallback_fee) {
@@ -2893,7 +2893,6 @@ bool CWallet::CreateTransactionInternal(
                     } else {
                         coin_selection_params.change_spend_size = (size_t)change_spend_size;
                     }
-                    coin_selection_params.effective_fee = nFeeRateNeeded;
                     if (!SelectCoins(vAvailableCoins, nValueToSelect, setCoins, nValueIn, coin_control, coin_selection_params, bnb_used))
                     {
                         // If BnB was used, it was the first pass. No longer the first pass and continue loop with knapsack.

--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2814,6 +2814,11 @@ bool CWallet::CreateTransactionInternal(
                 error = strprintf(_("Fee rate (%s) is lower than the minimum fee rate setting (%s)"), coin_control.m_feerate->ToString(FeeEstimateMode::SAT_VB), nFeeRateNeeded.ToString(FeeEstimateMode::SAT_VB));
                 return false;
             }
+            if (feeCalc.reason == FeeReason::FALLBACK && !m_allow_fallback_fee) {
+                // eventually allow a fallback fee
+                error = _("Fee estimation failed. Fallbackfee is disabled. Wait a few blocks or enable -fallbackfee.");
+                return false;
+            }
 
             nFeeRet = 0;
             bool pick_new_inputs = true;
@@ -2951,13 +2956,7 @@ bool CWallet::CreateTransactionInternal(
                     return false;
                 }
 
-                nFeeNeeded = GetMinimumFee(*this, nBytes, coin_control, &feeCalc);
-                if (feeCalc.reason == FeeReason::FALLBACK && !m_allow_fallback_fee) {
-                    // eventually allow a fallback fee
-                    error = _("Fee estimation failed. Fallbackfee is disabled. Wait a few blocks or enable -fallbackfee.");
-                    return false;
-                }
-
+                nFeeNeeded = coin_selection_params.effective_fee.GetFee(nBytes);
                 if (nFeeRet >= nFeeNeeded) {
                     // Reduce fee to only the needed amount if possible. This
                     // prevents potential overpayment in fees if the coins
@@ -2971,7 +2970,7 @@ bool CWallet::CreateTransactionInternal(
                     // change output. Only try this once.
                     if (nChangePosInOut == -1 && nSubtractFeeFromAmount == 0 && pick_new_inputs) {
                         unsigned int tx_size_with_change = nBytes + coin_selection_params.change_output_size + 2; // Add 2 as a buffer in case increasing # of outputs changes compact size
-                        CAmount fee_needed_with_change = GetMinimumFee(*this, tx_size_with_change, coin_control, nullptr);
+                        CAmount fee_needed_with_change = coin_selection_params.effective_fee.GetFee(tx_size_with_change);
                         CAmount minimum_value_for_change = GetDustThreshold(change_prototype_txout, discard_rate);
                         if (nFeeRet >= fee_needed_with_change + minimum_value_for_change) {
                             pick_new_inputs = false;

--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -2368,16 +2368,16 @@ bool CWallet::SelectCoinsMinConf(const CAmount& nTargetValue, const CoinEligibil
         // When subtracting the fee from the outputs, we want the effective feerate to be 0
         CFeeRate effective_feerate{0};
         if (!coin_selection_params.m_subtract_fee_outputs) {
-            effective_feerate = coin_selection_params.effective_fee;
+            effective_feerate = coin_selection_params.m_effective_feerate;
         }
 
         std::vector<OutputGroup> groups = GroupOutputs(coins, !coin_selection_params.m_avoid_partial_spends, effective_feerate, coin_selection_params.m_long_term_feerate, eligibility_filter, true /* positive_only */);
 
         // Calculate cost of change
-        CAmount cost_of_change = coin_selection_params.m_discard_feerate.GetFee(coin_selection_params.change_spend_size) + coin_selection_params.effective_fee.GetFee(coin_selection_params.change_output_size);
+        CAmount cost_of_change = coin_selection_params.m_discard_feerate.GetFee(coin_selection_params.change_spend_size) + coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.change_output_size);
 
         // Calculate the fees for things that aren't inputs
-        CAmount not_input_fees = coin_selection_params.effective_fee.GetFee(coin_selection_params.tx_noinputs_size);
+        CAmount not_input_fees = coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.tx_noinputs_size);
         bnb_used = true;
         return SelectCoinsBnB(groups, nTargetValue, cost_of_change, setCoinsRet, nValueRet, not_input_fees);
     } else {
@@ -2431,7 +2431,7 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
             if (coin.m_input_bytes <= 0) {
                 return false; // Not solvable, can't estimate size for fee
             }
-            coin.effective_value = coin.txout.nValue - coin_selection_params.effective_fee.GetFee(coin.m_input_bytes);
+            coin.effective_value = coin.txout.nValue - coin_selection_params.m_effective_feerate.GetFee(coin.m_input_bytes);
             if (coin_selection_params.use_bnb) {
                 value_to_select -= coin.effective_value;
             } else {
@@ -2802,11 +2802,11 @@ bool CWallet::CreateTransactionInternal(
             coin_selection_params.m_discard_feerate = GetDiscardRate(*this);
 
             // Get the fee rate to use effective values in coin selection
-            coin_selection_params.effective_fee = GetMinimumFeeRate(*this, coin_control, &feeCalc);
+            coin_selection_params.m_effective_feerate = GetMinimumFeeRate(*this, coin_control, &feeCalc);
             // Do not, ever, assume that it's fine to change the fee rate if the user has explicitly
             // provided one
-            if (coin_control.m_feerate && coin_selection_params.effective_fee > *coin_control.m_feerate) {
-                error = strprintf(_("Fee rate (%s) is lower than the minimum fee rate setting (%s)"), coin_control.m_feerate->ToString(FeeEstimateMode::SAT_VB), coin_selection_params.effective_fee.ToString(FeeEstimateMode::SAT_VB));
+            if (coin_control.m_feerate && coin_selection_params.m_effective_feerate > *coin_control.m_feerate) {
+                error = strprintf(_("Fee rate (%s) is lower than the minimum fee rate setting (%s)"), coin_control.m_feerate->ToString(FeeEstimateMode::SAT_VB), coin_selection_params.m_effective_feerate.ToString(FeeEstimateMode::SAT_VB));
                 return false;
             }
             if (feeCalc.reason == FeeReason::FALLBACK && !m_allow_fallback_fee) {
@@ -2955,7 +2955,7 @@ bool CWallet::CreateTransactionInternal(
                     return false;
                 }
 
-                nFeeNeeded = coin_selection_params.effective_fee.GetFee(nBytes);
+                nFeeNeeded = coin_selection_params.m_effective_feerate.GetFee(nBytes);
                 if (nFeeRet >= nFeeNeeded) {
                     // Reduce fee to only the needed amount if possible. This
                     // prevents potential overpayment in fees if the coins
@@ -2969,7 +2969,7 @@ bool CWallet::CreateTransactionInternal(
                     // change output. Only try this once.
                     if (nChangePosInOut == -1 && nSubtractFeeFromAmount == 0 && pick_new_inputs) {
                         unsigned int tx_size_with_change = nBytes + coin_selection_params.change_output_size + 2; // Add 2 as a buffer in case increasing # of outputs changes compact size
-                        CAmount fee_needed_with_change = coin_selection_params.effective_fee.GetFee(tx_size_with_change);
+                        CAmount fee_needed_with_change = coin_selection_params.m_effective_feerate.GetFee(tx_size_with_change);
                         CAmount minimum_value_for_change = GetDustThreshold(change_prototype_txout, coin_selection_params.m_discard_feerate);
                         if (nFeeRet >= fee_needed_with_change + minimum_value_for_change) {
                             pick_new_inputs = false;

--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -608,18 +608,20 @@ struct CoinSelectionParams
     size_t change_spend_size = 0;
     CFeeRate effective_fee = CFeeRate(0);
     CFeeRate m_long_term_feerate;
+    CFeeRate m_discard_feerate;
     size_t tx_noinputs_size = 0;
     //! Indicate that we are subtracting the fee from outputs
     bool m_subtract_fee_outputs = false;
     bool m_avoid_partial_spends = false;
 
     CoinSelectionParams(bool use_bnb, size_t change_output_size, size_t change_spend_size, CFeeRate effective_fee,
-                        CFeeRate long_term_feerate, size_t tx_noinputs_size, bool avoid_partial) :
+                        CFeeRate long_term_feerate, CFeeRate discard_feerate, size_t tx_noinputs_size, bool avoid_partial) :
         use_bnb(use_bnb),
         change_output_size(change_output_size),
         change_spend_size(change_spend_size),
         effective_fee(effective_fee),
         m_long_term_feerate(long_term_feerate),
+        m_discard_feerate(discard_feerate),
         tx_noinputs_size(tx_noinputs_size),
         m_avoid_partial_spends(avoid_partial)
     {}

--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -606,7 +606,7 @@ struct CoinSelectionParams
     bool use_bnb = true;
     size_t change_output_size = 0;
     size_t change_spend_size = 0;
-    CFeeRate effective_fee = CFeeRate(0);
+    CFeeRate m_effective_feerate;
     CFeeRate m_long_term_feerate;
     CFeeRate m_discard_feerate;
     size_t tx_noinputs_size = 0;
@@ -614,12 +614,12 @@ struct CoinSelectionParams
     bool m_subtract_fee_outputs = false;
     bool m_avoid_partial_spends = false;
 
-    CoinSelectionParams(bool use_bnb, size_t change_output_size, size_t change_spend_size, CFeeRate effective_fee,
+    CoinSelectionParams(bool use_bnb, size_t change_output_size, size_t change_spend_size, CFeeRate effective_feerate,
                         CFeeRate long_term_feerate, CFeeRate discard_feerate, size_t tx_noinputs_size, bool avoid_partial) :
         use_bnb(use_bnb),
         change_output_size(change_output_size),
         change_spend_size(change_spend_size),
-        effective_fee(effective_fee),
+        m_effective_feerate(effective_feerate),
         m_long_term_feerate(long_term_feerate),
         m_discard_feerate(discard_feerate),
         tx_noinputs_size(tx_noinputs_size),

--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -607,16 +607,19 @@ struct CoinSelectionParams
     size_t change_output_size = 0;
     size_t change_spend_size = 0;
     CFeeRate effective_fee = CFeeRate(0);
+    CFeeRate m_long_term_feerate;
     size_t tx_noinputs_size = 0;
     //! Indicate that we are subtracting the fee from outputs
     bool m_subtract_fee_outputs = false;
     bool m_avoid_partial_spends = false;
 
-    CoinSelectionParams(bool use_bnb, size_t change_output_size, size_t change_spend_size, CFeeRate effective_fee, size_t tx_noinputs_size, bool avoid_partial) :
+    CoinSelectionParams(bool use_bnb, size_t change_output_size, size_t change_spend_size, CFeeRate effective_fee,
+                        CFeeRate long_term_feerate, size_t tx_noinputs_size, bool avoid_partial) :
         use_bnb(use_bnb),
         change_output_size(change_output_size),
         change_spend_size(change_spend_size),
         effective_fee(effective_fee),
+        m_long_term_feerate(long_term_feerate),
         tx_noinputs_size(tx_noinputs_size),
         m_avoid_partial_spends(avoid_partial)
     {}

